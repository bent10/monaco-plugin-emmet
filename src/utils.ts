import expand, { extract } from 'emmet'
import * as monaco from 'monaco-editor'
import type { Range } from './types.js'

/**
 * Provides auto-completion suggestions for the given text model and
 * position using Emmet.
 *
 * @param model - The Monaco Editor text model.
 * @param position - The position in the text model where auto-completion is requested.
 * @returns The completion item or null if there are no suggestions.
 */
export function autoComplete(
  model: monaco.editor.ITextModel,
  position: monaco.Position
): monaco.languages.CompletionItem | void {
  const { lineNumber, column } = position
  const syntax = model.getLanguageId()

  const inlineText = model.getValueInRange({
    startLineNumber: 1,
    startColumn: 1,
    endLineNumber: lineNumber,
    endColumn: column
  })

  const { abbreviation = '', start = 1 } = extract(inlineText) || {}

  if (!abbreviation) return

  const range: Range = {
    startLineNumber: lineNumber,
    endLineNumber: lineNumber,
    startColumn: start,
    endColumn: column
  }

  const insertText = expand(abbreviation, {
    syntax,
    options: {
      'output.field': (index: string, placeholder: string) =>
        placeholder ? `\${${index}:${placeholder}}` : `\${${index}}`
    }
  })

  return {
    label: abbreviation,
    range,
    insertText,
    kind: monaco.languages.CompletionItemKind.Property,
    documentation: `Emmet ${abbreviation}`,
    insertTextRules:
      monaco.languages.CompletionItemInsertTextRule.InsertAsSnippet
  }
}
