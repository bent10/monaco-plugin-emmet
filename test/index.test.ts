import { registerEmmet, createSnippets } from '../src/index.js'
import { autoComplete } from '../src/utils.js'

type Api = Parameters<typeof registerEmmet>[0]
type Model = Parameters<typeof createSnippets>[0]
type Position = Parameters<typeof createSnippets>[1]

const mockMonaco = {
  languages: {
    registerCompletionItemProvider: vi.fn()
  }
} as unknown as Api

it('registers completion provider for each language', () => {
  registerEmmet(mockMonaco, ['html', 'css'], {
    html: {
      card: '.card>.card-body{${0}}'
    },
    css: {
      flex: 'display: flex;'
    }
  })

  expect(
    mockMonaco.languages.registerCompletionItemProvider
  ).toHaveBeenCalledTimes(2)

  expect(
    mockMonaco.languages.registerCompletionItemProvider
  ).toHaveBeenCalledWith('html', expect.any(Object))
  expect(
    mockMonaco.languages.registerCompletionItemProvider
  ).toHaveBeenCalledWith('css', expect.any(Object))
})

it('creates snippets for a given model, position, and snippets map', () => {
  const model = {
    getLanguageId: () => 'html',
    getWordUntilPosition: () => ({ startColumn: 1, endColumn: 4 })
  } as unknown as Model
  const position = { lineNumber: 1, column: 3 } as Position
  const snippetsMap = {
    card: '.card>.card-body{${1}}',
    'btn:primary': 'button[class="${1:btn btn-primary}"]{${2}}'
  }

  const snippets = createSnippets(model, position, snippetsMap)

  expect(snippets).toHaveLength(2)
  expect(snippets[0].label).toBe('card')
  expect(snippets[1].label).toBe('btn:primary')
  expect(snippets[0].insertText).toMatchSnapshot()
  expect(snippets[1].insertText).toMatchSnapshot()
})

it('returns null for autoComplete if no abbreviation is found', () => {
  const model = {
    getValueInRange: () => '',
    getLanguageId: () => 'html'
  } as unknown as Model
  const position = { lineNumber: 1, column: 5 } as Position
  const result = autoComplete(model, position)
  expect(result).toBeNull()
})

it('returns autoComplete suggestion for the given model and position', () => {
  const model = {
    getValueInRange: () => '<div',
    getLanguageId: () => 'html'
  } as unknown as Model
  const position = { lineNumber: 1, column: 3 } as Position
  const result = autoComplete(model, position)
  expect(result).toBeDefined()
  expect(result!.label).toBe('div')
  expect(result!.insertText).toBe('<div>${1}</div>')
  expect(result!.range).toEqual({
    startLineNumber: 1,
    endLineNumber: 1,
    startColumn: 1,
    endColumn: 3
  })
})

it('expands user-defined snippets with placeholders', () => {
  const model = {
    getLanguageId: () => 'html',
    getWordUntilPosition: () => ({ startColumn: 1, endColumn: 4 })
  } as unknown as Model
  const position = { lineNumber: 1, column: 3 } as Position
  const snippetsMap = {
    card: '.card>.card-body{${1:lorem}}'
  }
  const snippets = createSnippets(model, position, snippetsMap)
  expect(snippets[0].insertText).toMatchSnapshot()
})

// Edge case: Test expansion with no placeholders
it('expands user-defined snippets with no placeholders', () => {
  const model = {
    getLanguageId: () => 'html',
    getWordUntilPosition: () => ({ startColumn: 1, endColumn: 4 })
  } as unknown as Model
  const position = { lineNumber: 1, column: 3 } as Position
  const snippetsMap = {
    card: '.card>.card-body'
  }
  const snippets = createSnippets(model, position, snippetsMap)
  expect(snippets[0].insertText).toMatchSnapshot()
})

// Edge case: Test expansion with complex placeholders
it('expands user-defined snippets with complex placeholders', () => {
  const model = {
    getLanguageId: () => 'html',
    getWordUntilPosition: () => ({ startColumn: 1, endColumn: 4 })
  } as unknown as Model
  const position = { lineNumber: 1, column: 3 } as Position
  const snippetsMap = {
    card: '.card>.card-body{${1:btn} ${2}}'
  }
  const snippets = createSnippets(model, position, snippetsMap)
  expect(snippets[0].insertText).toMatchSnapshot()
})

// Edge case: Test empty snippets map
it('does not create snippets for an empty snippets map', () => {
  const model = {
    getLanguageId: () => 'html',
    getWordUntilPosition: () => ({ startColumn: 1, endColumn: 4 })
  } as unknown as Model
  const position = { lineNumber: 1, column: 3 } as Position
  const snippetsMap = {}
  const snippets = createSnippets(model, position, snippetsMap)
  expect(snippets).toHaveLength(0)
})
