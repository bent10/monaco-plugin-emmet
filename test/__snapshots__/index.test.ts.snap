// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`creates snippets for a given model, position, and snippets map 1`] = `
"<div class="card">
	<div class="card-body">\${2}</div>
</div>"
`;

exports[`creates snippets for a given model, position, and snippets map 2`] = `"<button class="\${2:btn btn-primary}">\${5}</button>"`;

exports[`expands user-defined snippets with complex placeholders 1`] = `
"<div class="card">
	<div class="card-body">\${2:btn} \${3}</div>
</div>"
`;

exports[`expands user-defined snippets with no placeholders 1`] = `
"<div class="card">
	<div class="card-body">\${1}</div>
</div>"
`;

exports[`expands user-defined snippets with placeholders 1`] = `
"<div class="card">
	<div class="card-body">\${2:lorem}</div>
</div>"
`;
